tests/test_linearity.py

```diff
@@ -0,0 +1,407 @@
+            pd_handles.append(
+                lsst.pipe.base.InMemoryDatasetHandle(
+                    pd_calib,
+                    dataId={"exposure": exp_id_pair[1]},
+                )
+            )
+
+        config = LinearitySolveTask.ConfigClass()
+        config.linearityType = "Spline"
+        config.usePhotodiode = True
+        config.photodiodeIntegrationMethod = "CHARGE_SUM"
+        config.minLinearAdu = 0.0
+        config.maxLinearAdu = np.max(mu_values) + 1.0
+        config.splineKnots = n_nodes
+
+        if do_pd_offsets:
+            config.splineGroupingColumn = "CCOBCURR"
+
+        task = LinearitySolveTask(config=config)
+        linearizer = task.run(
+            ptc,
+            [self.dummy_exposure],
+            self.camera,
+            self.input_dims,
+            inputPhotodiodeData=pd_handles,
+        ).outputLinearizer
+
+        # Skip the last amp which is marked bad.
+        for amp_name in ptc.ampNames[:-1]:
+            lin_mask = np.isfinite(linearizer.fitResiduals[amp_name])
+
+            # Make sure that anything in the input mask is still masked.
+            check, = np.where(~ptc.expIdMask[amp_name])
+            if len(check) > 0:
+                self.assertEqual(np.all(lin_mask[check]), False)
+
+            # Make sure the outliers are masked.
+            self.assertEqual(np.all(lin_mask[outlier_indices]), False)
+
+            # The first point at very low flux is noisier and so we exclude
+            # it from the test here.
+            self.assertFloatsAlmostEqual(
+                (linearizer.fitResiduals[amp_name][lin_mask] / mu_linear[lin_mask])[1:],
+                0.0,
+                atol=1e-3,
+            )
+
+            # If we apply the linearity correction, we should get the true
+            # linear values out.
+            image = lsst.afw.image.ImageF(len(mu_values), 1)
+            image.array[:, :] = mu_values
+            lin_func = linearizer.getLinearityTypeByName(linearizer.linearityType[amp_name])
+            lin_func()(
+                image,
+                coeffs=linearizer.linearityCoeffs[amp_name],
+                log=None,
+            )
+
+            # We scale by the median because of ambiguity in the overall
+            # gain parameter which is not part of the non-linearity.
+            ratio = image.array[0, lin_mask]/mu_linear[lin_mask]
+            self.assertFloatsAlmostEqual(
+                ratio / np.median(ratio),
+                1.0,
+                rtol=5e-4,
+            )
+
+            # Check that the spline parameters recovered are consistent,
+            # with input to some low-grade precision.
+            # The first element should be identically zero.
+            self.assertFloatsEqual(linearizer.linearityCoeffs[amp_name][0], 0.0)
+
+            self.assertFloatsAlmostEqual(
+                linearizer.linearityCoeffs[amp_name][n_nodes:],
+                non_lin_spline_values,
+                atol=7.0,
```

Comment:
Is this tolerance needed for all of the spline parameters?  I'm less concerned about a 70/7 or 185/7 S/N, but a 1/7 would be concerning.

Reply:
It's not, it's just at the large ones.  But the ratio comparison wasn't looking good.  The proof is in the linearity of the residuals, but I wanted to add some sort of test here.  If you think it's better to do different tolerances at low and high flux I can do that instead.

Reply:
Ok.  If my specific worry isn't justified, then this is fine with me.